app/src/test/java/com/github/onlynotesswent/model/authentication/AuthenticatorTest.kt

```diff
@@ -0,0 +1,428 @@
+      authenticator.handleSignIn(getCredentialResponse, {}, { onFailureCalled = true })
+
+      assert(onFailureCalled)
+
+      verifyErrorLog("Received an invalid google id token response")
+    }
+  }
+
+  @Test
+  fun `handleSignIn should log error when invalid custom credential type`() = runTest {
+    `when`(mockCustomCredential.type).thenReturn("INVALID_TYPE")
+    var onFailureCalled = false
+    authenticator.handleSignIn(mockGetCredentialResponse, {}, { onFailureCalled = true })
+
+    assert(onFailureCalled)
+
+    verifyErrorLog("Unexpected type of credential: INVALID_TYPE")
+  }
+
+  @Test
+  fun `handleSignIn should log error when invalid credential type`() = runTest {
+    val mockCredential = mock<Credential>()
+    `when`(mockGetCredentialResponse.credential).thenReturn(mockCredential)
+    // `when`(mockCredential.javaClass).thenReturn(Credential::class.java)
+
+    `when`(mockCredential.type).thenReturn("INVALID_TYPE")
+    var onFailureCalled = false
+    authenticator.handleSignIn(mockGetCredentialResponse, {}, { onFailureCalled = true })
+
+    assert(onFailureCalled)
+
+    verifyErrorLog("Unexpected type of credential")
+  }
+
+  //  @Test
+  //  fun `handleSignIn should log error signing in with Firebase`() = runTest {
+  //    //var onFailureCalled = false
+  //    //authenticator.handleSignIn(mockGetCredentialResponse, {}, { onFailureCalled = true })
+  //
+  //    val MockGoogleIdTokenCredential = (GoogleIdTokenCredential::class.java)
+  //    val mockGoogleIdTokenCredential = mock<GoogleIdTokenCredential>()
+  //    MockGoogleIdTokenCredential.`when`<GoogleIdTokenCredential> {
+  // GoogleIdTokenCredential.createFrom(any()) }
+  //        .thenReturn(mockGoogleIdTokenCredential)
+  //
+  ////    val MockGoogleAuthProvider = mockStatic(GoogleIdTokenCredential::class.java)
+  ////    val mockGoogleIdTokenCredential = mock<GoogleIdTokenCredential>()
+  ////    MockGoogleIdTokenCredential.`when`<GoogleIdTokenCredential> {
+  // GoogleIdTokenCredential.createFrom(any()) }
+  ////      .thenReturn(mockGoogleIdTokenCredential)
+  //
+  //    //assert(onFailureCalled)
+  //
+  //    verifyErrorLog("Received an invalid google id token response")
+  //  }
+
+  @Test
+  fun `generateNonce should generate a hashed nonce`() {
+    // Act
+    val nonce1 = generateNonce()
+    val nonce2 = generateNonce()
+
+    // Assert
+    assertNotNull(nonce1)
+    assertNotNull(nonce2)
+    assertNotEquals(nonce1, nonce2)
+    assertEquals(64, nonce1.length) // SHA-256 hash length in hex is 64 characters
+    assertEquals(64, nonce2.length)
+  }
+}
+
+private fun verifyErrorLog(msg: String) {
+  // Get all the logs
+  val logs = ShadowLog.getLogs()
+
+  // Check for the debug log that should be generated
+  val errorLog = logs.find { it.type == Log.ERROR && it.tag == Authenticator.TAG && it.msg == msg }
+  assert(errorLog != null) { "Expected error log was not found!" }
+}
+
+// package com.github.onlynotesswent.model.authentication
```

Comment:
As mentioned above, testing sign in is difficult. Coverage is good enough, tests not used and commented out could be removed